crates/polars-core/src/chunked_array/logical/struct_/mod.rs

```diff
@@ -173,58 +177,54 @@ impl StructChunked {
-        for i in 0..chunks_lens {
-            for s in self.fields() {
-                let arr = &s.chunks()[i];
-                let has_nulls = arr.null_count() > 0 || matches!(s.dtype(), DataType::Null);
-                if !has_nulls {
-                    self.null_count = 0;
-                    return;
-                }
-            }
+        // Count both the total number of nulls and the rows where everything is null
+        (self.null_count, self.total_null_count) = (0, 0);
+
+        // If there is at least one field with no null values, no rows are null. However, we still
+        // have to count the number of nulls per field to get the total number. Fortunately this is
+        // cheap since null counts per chunk are pre-computed.
+        let (could_have_nulls, total_null_count) =
+            self.fields().iter().fold((true, 0), |acc, s| {
+                (acc.0 & (s.null_count() != 0), acc.1 + s.null_count())
+            });
+        self.total_null_count = total_null_count;
+        if !could_have_nulls {
+            return;
         }
-
-        // slow path
-        // we bitand every null validity bitmask to determine
-        // in which rows all values are null
-        for i in 0..chunks_lens {
-            let mut validity_agg = None;
-
-            let mut all_null_array = true;
+        // A row is null if all values in it are null, so we bitor every validity bitmask since a
+        // single valid entry makes that row not null. We can also save some work by not bothering
+        // to bitor fields that would have all 0 validities (Null dtype or everything null).
+        for i in 0..self.fields()[0].chunks().len() {
+            let mut validity_agg: Option<arrow::bitmap::Bitmap> = None;
+            let mut n_nulls = None;
             for s in self.fields() {
                 let arr = &s.chunks()[i];
-
-                if !matches!(s.dtype(), DataType::Null) {
-                    all_null_array = false;
-                    match (&validity_agg, arr.validity()) {
-                        (Some(agg), Some(validity)) => validity_agg = Some(validity.bitand(agg)),
-                        (None, Some(validity)) => validity_agg = Some(validity.clone()),
-                        _ => {}
+                if s.dtype() == &DataType::Null {
+                    // The implicit validity mask is all 0 so it wouldn't affect the bitor
+                    continue;
+                }
+                match (arr.validity(), n_nulls, arr.null_count() == 0) {
+                    // The null count is to avoid touching chunks with a validity mask but no nulls
+                    (_, Some(0), _) => break, // No all-null rows, next chunk!
+                    (None, _, _) | (_, _, true) => n_nulls = Some(0),
+                    (Some(v), _, _) => {
+                        validity_agg =
+                            validity_agg.map_or(Some(v.clone()), |agg| Some(v.bitor(&agg)));
```

Comment:
`map_or` is eagerly evaluated. We can use  `map_or_else` here to don't do the clone until we need. it.